pkg/dns/aws/dns.go

```diff
@@ -0,0 +1,248 @@
+		return nil
+	}
+
+	// Find and cache all zones and their tags.
+	// TODO: probably only need to store the two zones we care about.
+	zones := []*zoneWithTags{}
+	listedAllTags := true
+	f := func(resp *route53.ListHostedZonesOutput, lastPage bool) (shouldContinue bool) {
+		for _, zone := range resp.HostedZones {
+			tagsList, err := m.Route53.ListTagsForResource(&route53.ListTagsForResourceInput{
+				ResourceType: aws.String("hostedzone"),
+				ResourceId:   zone.Id,
+			})
+			if err != nil {
+				logrus.Errorf("failed to list tags for hosted zone %s: %v", zone.Id, err)
+				listedAllTags = false
+				return false
+			}
+			zones = append(zones, &zoneWithTags{Zone: zone, Tags: tagsList.ResourceTagSet.Tags})
+		}
+		return true
+	}
+	if !listedAllTags {
+		return fmt.Errorf("failed to list all hosted zone tags")
+	}
+	err := m.Route53.ListHostedZonesPages(&route53.ListHostedZonesInput{}, f)
+	if err != nil {
+		return fmt.Errorf("failed to list hosted zones: %v", err)
+	}
+
+	m.zones = zones
+	return nil
+}
+
+// findClusterPublicZone finds the public zone given the base domain.
+func (m *Manager) findClusterPublicZone() *route53.HostedZone {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	for _, zwt := range m.zones {
+		if aws.StringValue(zwt.Zone.Name) == m.config.BaseDomain && !aws.BoolValue(zwt.Zone.Config.PrivateZone) {
+			return zwt.Zone
+		}
+	}
+	return nil
+}
+
+// findClusterPrivateZone finds the private zone given the base domain and
+// cluster ID.
+func (m *Manager) findClusterPrivateZone() *route53.HostedZone {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	for _, zwt := range m.zones {
+		if aws.StringValue(zwt.Zone.Name) == m.config.BaseDomain && aws.BoolValue(zwt.Zone.Config.PrivateZone) {
+			for _, tag := range zwt.Tags {
+				if aws.StringValue(tag.Key) == "tectonicClusterID" && aws.StringValue(tag.Value) == m.config.ClusterID {
+					return zwt.Zone
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// EnsureAlias will create an A record for domain/target in both the public and
+// private zones. The target must correspond to the hostname of an ELB which
+// will be automatically discovered.
+func (m *Manager) EnsureAlias(domain, target string) error {
+	if len(domain) == 0 {
+		return fmt.Errorf("domain is required")
+	}
+	if len(target) == 0 {
+		return fmt.Errorf("target is required")
+	}
+
+	// Ensure the zone cache is up to date.
+	err := m.updateZoneCache()
+	if err != nil {
+		return fmt.Errorf("failed to update hosted zone cache: %v", err)
+	}
+
+	publicZone := m.findClusterPublicZone()
+	privateZone := m.findClusterPrivateZone()
+	if publicZone == nil {
+		return fmt.Errorf("failed to find public zone for cluster")
+	}
+	if privateZone == nil {
+		return fmt.Errorf("failed to find private zone for cluster")
+	}
+
+	// Find the target hosted zone of the load balancer attached to the service.
+	// TODO: cache it?
+	var targetHostedZoneID string
+	loadBalancers, err := m.ELB.DescribeLoadBalancers(&elb.DescribeLoadBalancersInput{})
```

Comment:
You'll need the elbv2 version of this call to catch the new network load balancers we're using

Reply:
Can you explain further? I'm having no issues w/ the ELBs created by the k8s cloud provider for our services.